File: py/configu/commands/export_command.py
import os
from typing import Dict, TypedDict

from .eval_command import EvalCommandReturn
from ..core import (
    Command,
)


class ExportCommandParameters(TypedDict):
    pipe: EvalCommandReturn
    env: bool
    override: bool


ExportCommandReturn = Dict[str, str]


class ExportCommand(Command[ExportCommandReturn]):
    """"""

    parameters: ExportCommandParameters

    def __init__(
        self,
        pipe: EvalCommandReturn,
        *,
        env: bool = True,
        override: bool = True,
    ) -> None:
        super().__init__(ExportCommandParameters(pipe=pipe, env=env, override=override))

    def run(self) -> ExportCommandReturn:
        env = self.parameters.get("env", True)
        override = self.parameters.get("override", True)
        pipe = self.parameters["pipe"]
        exported_configs: ExportCommandReturn = {
            key: value["result"]["value"] for key, value in pipe.items()
        }
        if env:
            for key, value in exported_configs.items():
                if override or key not in os.environ:
                    os.environ[key] = value

        return exported_configs

File: packages/ts/src/commands/ExportCommand.ts
import _ from 'lodash';
import { Command } from '../Command';
import { type EvalCommandReturn } from './EvalCommand';
import { NAME, ConfigError } from '../utils';

export type ExportCommandReturn = {
  [key: string]: string;
};

export type ExportCommandParameters = {
  pipe: EvalCommandReturn;
  keys?: (key: string) => string;
  filter?: ({ context, result }: EvalCommandReturn['string']) => boolean;
};

export class ExportCommand extends Command<ExportCommandReturn> {
  constructor(public parameters: ExportCommandParameters) {
    super(parameters);
  }

  private mutateKeys(result: { [key: string]: string }): { [key: string]: string } {
    const { keys } = this.parameters;

    if (!keys) {
      return result;
    }

    return _.mapKeys(result, (current, key) => {
      const errorScope: [string, string][] = [
        ['ExportCommand', ''],
        ['ConfigKey', key],
      ];
      const mutatedKey = String(keys(key));
      if (!NAME(mutatedKey)) {
        throw new ConfigError('invalid config key', `key "${key}" mustn't contain reserved words`, errorScope);
      }
      return mutatedKey;
    });
  }

  private filterPipe(pipe: EvalCommandReturn) {
    const { filter } = this.parameters;
    if (!filter) {
      return _.pickBy(pipe, ({ context }) => !context.cfgu.hidden);
    }
    return _.pickBy(pipe, filter);
  }

  async run() {
    const { pipe } = this.parameters;
    const filteredPipe = this.filterPipe(pipe);
    const configDict = _.mapValues(filteredPipe, (current) => current.result.value);
    const keyMutatedConfigDict = this.mutateKeys(configDict);
    return keyMutatedConfigDict;
  }
}
